chromium_src/ios/chrome/browser/shared/model/prefs/browser_prefs.mm

```diff
@@ -3,85 +3,21 @@
-#include "brave/components/p3a/rotation_scheduler.h"
-#include "brave/components/skus/browser/skus_utils.h"
-#include "brave/ios/browser/brave_stats/brave_stats_prefs.h"
 #include "components/pref_registry/pref_registry_syncable.h"
-#include "ios/chrome/browser/shared/model/prefs/pref_names.h"
+#include "components/prefs/pref_service.h"
 
-void BraveRegisterBrowserStatePrefs(
-    user_prefs::PrefRegistrySyncable* registry) {
-  brave_ads::RegisterProfilePrefs(registry);
-  brave_ads::RegisterProfilePrefsForMigration(registry);
-  brave_rewards::RegisterProfilePrefs(registry);
-  brave_rewards::RegisterProfilePrefsForMigration(registry);
-  brave_sync::Prefs::RegisterProfilePrefs(registry);
-  brave_wallet::RegisterProfilePrefs(registry);
-  brave_wallet::RegisterProfilePrefsForMigration(registry);
-  de_amp::RegisterProfilePrefs(registry);
-  debounce::DebounceService::RegisterProfilePrefs(registry);
-  ai_chat::prefs::RegisterProfilePrefs(registry);
-  ai_chat::ModelService::RegisterProfilePrefs(registry);
-  omnibox::RegisterBraveProfilePrefs(registry);
-  brave_news::prefs::RegisterProfilePrefs(registry);
-  ntp_background_images::RegisterProfilePrefs(registry);
-  ntp_background_images::RegisterProfilePrefsForMigration(registry);
-}
-
-void BraveRegisterLocalStatePrefs(PrefRegistrySimple* registry) {
-  brave_stats::RegisterLocalStatePrefs(registry);
-  brave_wallet::RegisterLocalStatePrefs(registry);
-  brave_wallet::RegisterLocalStatePrefsForMigration(registry);
-  decentralized_dns::RegisterLocalStatePrefs(registry);
-  skus::RegisterLocalStatePrefs(registry);
-  p3a::P3AService::RegisterPrefs(registry, false);
-  p3a::MetricLogStore::RegisterLocalStatePrefsForMigration(registry);
-  p3a::RotationScheduler::RegisterLocalStatePrefsForMigration(registry);
-  ntp_background_images::NTPBackgroundImagesService::RegisterLocalStatePrefs(
-      registry);
-  brave_l10n::RegisterLocalStatePrefsForMigration(registry);
-  ai_chat::prefs::RegisterLocalStatePrefs(registry);
-  ai_chat::AIChatMetrics::RegisterPrefs(registry);
-  ntp_background_images::RegisterLocalStatePrefs(registry);
-
-  // brave_shields
-  // Note this can be removed when we use the entire
-  // `brave_shields:AdBlockService` And we can call its
-  // `RegisterPrefsForAdBlockService`
-  registry->RegisterDictionaryPref(
-      brave_shields::prefs::kAdBlockRegionalFilters);
-  registry->RegisterDictionaryPref(
-      brave_shields::prefs::kAdBlockListSubscriptions);
-  registry->RegisterBooleanPref(
-      brave_shields::prefs::kAdBlockCheckedAllDefaultRegions, false);
-  registry->RegisterBooleanPref(
-      brave_shields::prefs::kAdBlockCheckedDefaultRegion, false);
-}
+namespace brave {
+void RegisterBrowserStatePrefs(user_prefs::PrefRegistrySyncable* registry);
+void RegisterLocalStatePrefs(PrefRegistrySimple* registry);
+void MigrateObsoleteProfilePrefs(PrefService* prefs);
+void MigrateObsoleteLocalStatePrefs(PrefService* prefs);
+}  // namespace brave
```

Comment:
I'm confused. why aren't these declared in a header and get included?

Reply:
Its a new pattern for overrides where we declare the functions in the override and implement the symbols on the brave side instead. The methods themselves are only called by the override so there's no header on the Brave side

Reply:
This works well as long as the Brave side of the code is always compile alongside the Chromium one

Reply:
but how do we make sure there is no missing direct deps? Being able to compile could also mean it can be using transitive deps which makes build unstable.

Reply:
I would expect only the linking stage to fail with undefined symbols error if we didn't compile the Brave side sources, and linking happens last so should be fine. cc @bridiver in case I'm missing something